ssl/ssl_sess.c

```diff
@@ -59,17 +59,37 @@ __owur static int timeoutcmp(SSL_SESSION *a, SSL_SESSION *b)
  */
 void ssl_session_calculate_timeout(SSL_SESSION *ss)
 {
-#ifndef __DJGPP__ /* time_t is unsigned on djgpp */
-    /* Force positive timeout */
-    if (ss->timeout < 0)
-        ss->timeout = 0;
+    time_t tmax = -1;
+
+    if (sizeof(time_t) == 8) {
+        uint64_t overflow;
+
+#ifndef __DJGPP__ /* time_t is unsigned on djgpp, it's signed anywhere else */
+        tmax = (time_t)((uint64_t)tmax >> 1);
 #endif
-    ss->calc_timeout = ss->time + ss->timeout;
-    /*
-     * |timeout| is always zero or positive, so the check for
-     * overflow only needs to consider if |time| is positive
-     */
-    ss->timeout_ovf = ss->time > 0 && ss->calc_timeout < ss->time;
+        overflow = (uint64_t)((uint64_t)tmax - ss->time);
+        if (ss->timeout > (time_t)overflow) {
+            ss->timeout_ovf |= 1;
+            ss->calc_timeout = ss->timeout - overflow;
+        } else {
+            ss->calc_timeout = ss->time + ss->timeout;
+        }
+    } else {
+        uint32_t overflow;
+
+#ifndef __DJGPP__ /* time_t is unsigned on djgp, it's signed anywhere elsep */
+        tmax = (time_t)((uint32_t)tmax >> 1);
+#endif
+        overflow = (uint32_t)((uint32_t)tmax - ss->time);
```

Comment:
Same here.

Reply:
the case at line 83 is needed to keep win32 build happy. this is the error log:
```
..\ssl\ssl_sess.c(83): error C2220: the following warning is treated as an error
..\ssl\ssl_sess.c(83): warning C4244: '=': conversion from 'time_t' to 'uint32_t', possible loss of data
NMAKE : fatal error U1077: '"cl"  /Zi /Fdossl_static.pdb /Gs0 /GF /Gy /MD /W3 /wd4090 /nologo /O2 /WX -I"." -I"include" -I".." -I"..\include" -D"L_ENDIAN" -D"OPENSSL_PIC" -D"OPENSSLDIR=\"C:\\Program Files (x86)\\Common Files\\SSL\"" -D"ENGINESDIR=\"C:\\Program Files (x86)\\OpenSSL\\lib\\engines-3\"" -D"MODULESDIR=\"C:\\Program Files (x86)\\OpenSSL\\lib\\ossl-modules\"" -D"OPENSSL_BUILDING_OPENSSL" -D"OPENSSL_SYS_WIN32" -D"WIN32_LEAN_AND_MEAN" -D"UNICODE" -D"_UNICODE" -D"_CRT_SECURE_NO_DEPRECATE" -D"_WINSOCK_DEPRECATED_NO_WARNINGS" -D"NDEBUG"   -DAES_ASM -c /Fossl\libssl-shlib-ssl_sess.obj "..\ssl\ssl_sess.c"' : return code '0x2'
Stop.
NMAKE : fatal error U1077: '"C:\Program Files\Microsoft Visual Studio\2022\Enterprise\VC\Tools\MSVC\14.40.33807\bin\HostX86\x86\nmake.exe" /S                  _build_sw' : return code '0x2'
Stop.
```
full log here.
https://github.com/openssl/openssl/actions/runs/10485823813/job/29042748003

this is what makes the compiler grumpy:
```
overflow = ((uint32_t)tmax - ss->time);
```
this is what makes the compiler happy
```
overflow = (uint32_t)((uint32_t)tmax - ss->time);
```
the happy fix looks odd, but seems to work. may be it should be
```
overflow = (uint32_t)((uint32_t)tmax - ss->time);
```